pageserver/src/tenant/timeline/import_pgdata.rs

```diff
@@ -33,166 +44,214 @@ pub async fn doit(
-
-        flow::run(timeline.clone(), control_file, storage.clone(), ctx).await?;
-
-        //
-        // Communicate that shard is done.
-        // Ensure at-least-once delivery of the upcall to storage controller
-        // before we mark the task as done and never come here again.
-        //
-        let storcon_client = StorageControllerUpcallClient::new(timeline.conf, &cancel);
-        storcon_client
-            .put_timeline_import_status(
-                timeline.tenant_shard_id,
-                timeline.timeline_id,
-                // TODO(vlad): What about import errors?
-                ShardImportStatus::Done,
-            )
-            .await
-            .map_err(|_err| anyhow::anyhow!("Shut down while putting timeline import status"))?;
-
-        storage
-            .put_json(
-                &shard_status_key,
-                &importbucket_format::ShardStatus { done: true },
-            )
-            .await
-            .context("put shard status")?;
-    }
+        Some(ShardImportStatus::Finalizing) => {
+            //
+            // Mark as done in index_part.
+            // This makes subsequent timeline loads enter the normal load code path
+            // instead of spawning the import task and calling this here function.
+            //
+            info!("mark import as complete in index part");
+            timeline
+                .remote_client
+                .schedule_index_upload_for_import_pgdata_state_update(Some(
+                    index_part_format::Root::V1(index_part_format::V1::Done(
+                        index_part_format::Done {
+                            idempotency_key,
+                            started_at,
+                            finished_at: chrono::Utc::now().naive_utc(),
+                        },
+                    )),
+                ))?;
+
+            timeline.remote_client.wait_completion().await?;
 
-    //
-    // Mark as done in index_part.
-    // This makes subsequent timeline loads enter the normal load code path
-    // instead of spawning the import task and calling this here function.
-    //
-    info!("mark import as complete in index part");
-    timeline
-        .remote_client
-        .schedule_index_upload_for_import_pgdata_state_update(Some(index_part_format::Root::V1(
-            index_part_format::V1::Done(index_part_format::Done {
-                idempotency_key,
-                started_at,
-                finished_at: chrono::Utc::now().naive_utc(),
-            }),
-        )))?;
-
-    timeline.remote_client.wait_completion().await?;
+            storcon_client
+                .put_timeline_import_status(
+                    timeline.tenant_shard_id,
+                    timeline.timeline_id,
+                    ShardImportStatus::Done,
+                )
+                .await
+                .map_err(|_err| {
+                    anyhow::anyhow!("Shut down while putting timeline import status")
+                })?;
```

Comment:
smelly that we're duplicating this snippet of code from right above. Why can't this be shared?

Reply:
I removed the finalizing state in https://github.com/neondatabase/neon/pull/11855/commits/e50f8e9815ffb2bbb3dc91c35b4ac84ecb7fd944, so the duplication is gone.